solver/bboltcachestorage/storage.go

```diff
@@ -455,3 +460,50 @@ func isEmptyBucket(b *bolt.Bucket) bool {
 	k, _ := b.Cursor().First()
 	return k == nil
 }
+
+// safeOpenDB opens a bolt database and recovers from panic that
+// can be caused by a corrupted database file.
+func safeOpenDB(dbPath string) (db *bolt.DB, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf("%v", r)
+		}
+
+		// If we get an error when opening the database, but we have
+		// access to the file, then fallback to resetting the database
+		// since the database may be corrupt.
+		if err != nil && canAccessFile(dbPath) {
+			db, err = fallbackOpenDB(dbPath, err)
+		}
+	}()
+	return openDB(dbPath)
+}
+
+// fallbackOpenDB performs the panic recovery, database backup, and
```

Comment:
"performs database backup, ... . Called after first db.Open has failed on reading the DB file."